utilities/agg_merge/agg_merge.cc

```diff
@@ -0,0 +1,206 @@
+                                  const Slice& value) {
+  std::string result;
+  PutLengthPrefixedSlice(&result, function_name);
+  result += value.ToString();
+  return result;
+}
+
+bool ExtractAggFuncAndValue(const Slice& op, Slice& func, Slice& value) {
+  value = op;
+  return GetLengthPrefixedSlice(&value, &func);
+}
+
+bool ExtractList(const Slice& encoded_list, std::vector<Slice>& decoded_list) {
+  decoded_list.clear();
+  Slice list_slice = encoded_list;
+  bool ret;
+  Slice item;
+  while ((ret = GetLengthPrefixedSlice(&list_slice, &item))) {
+    decoded_list.push_back(item);
+  }
+  return list_slice.empty();
+}
+
+class Accumulator {
+ public:
+  bool Add(const Slice& op, bool is_initial_value,
+           bool is_partial_aggregation) {
+    if (ignore_operands_) {
+      return true;
+    }
+    Slice my_func;
+    Slice my_value;
+    bool ret = ExtractAggFuncAndValue(op, my_func, my_value);
+    if (!ret || (!func_.empty() && func_ != my_func)) {
+      if (is_initial_value) {
+        // We allow initial value not to have the function name part.
+        my_value = op;
+        my_func = func_;
+      } else {
+        // We got some unexpected merge operands. Ignore this and all
+        // subsequence ones.
+        ignore_operands_ = true;
+        return true;
+      }
+    }
+    if (is_partial_aggregation && !func_.empty()) {
+      auto f = func_map.find(func_.ToString());
+      if (f == func_map.end() || f->second->DoPartialAggregate()) {
+        return false;
+      }
+    }
+    func_ = my_func;
+    values_.push_back(my_value);
+    return true;
+  }
+
+  // Return false if aggregation fails.
+  // One possible reason
+  bool GetResult(std::string* result) {
+    if (func_.empty()) {
+      return false;
+    }
+    auto f = func_map.find(func_.ToString());
+    if (f == func_map.end()) {
+      return false;
+    }
+    if (!f->second->Aggregate(values_, &scratch_)) {
+      return false;
+    }
+    *result = EncodeAggFuncAndValue(func_, scratch_);
+    return true;
+  }
+
+  void Clear() {
+    func_.clear();
+    values_.clear();
+    scratch_.clear();
+    ignore_operands_ = false;
+  }
+
+ private:
+  Slice func_;
+  std::vector<Slice> values_;
+  std::string scratch_;
+  bool ignore_operands_ = false;
+};
+
+// Creating and using a new Accumulator might invoke multiple malloc and is
+// expensive if it needs to be done when processing each merge operation.
+// AggMergeOperator's merge operators can be invoked concurrently by multiple
+// threads so we cannot simply create one Aggregator and reuse.
+// We use thread local instances instead.
+Accumulator& GetTLSAccumulator() {
+  // The implementation is mostly copoied from Random::GetTLSInstance()
+  // If the same pattern is used more frequently, we might create a utility
+  // function for that.
+  static __thread Accumulator* tls_instance;
+  static __thread std::aligned_storage<sizeof(Accumulator)>::type
+      tls_instance_bytes;
```

Comment:
It is a longer discussion. I do hope that now all platform we are supporting can use thread_local so that we can clear them all. My teammates don't feel confidence about it yet, so we might need later to find time to clear it up. The practice we are following it in perf_context_impl.h's definition of perf_context. Here we are following a simpler example of using __thread. I think we need to clear them all but I am right now following the existing practice.

Reply:
C++ has had thread_local support since C++11 and RocksDB is using C++17 now, so thread_local should not be a concern.  There are other PRs/issues to address/remove __thread, which is not portable.  And some of the existing code already does that (from port/port_posix.h):
// thread_local is part of C++11 and later (TODO: clean up this define)
#ifndef __thread
#define __thread thread_local
#endif